From 047c8f3ea16fe1d071f83b51b8c132f797903c91 Mon Sep 17 00:00:00 2001
From: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
Date: Tue, 10 Oct 2023 09:13:38 +0530
Subject: [PATCH] wifi: ath11k: Add support to send the QoS Null Data frame
 through exception path

When we try to send QoS NULL Data frame in Ethernet encap type, it modified
as QoS Data frame with TID 0 (encryption enabled one if security enabled).
But expectation is, it should be send as open type frame with TID 7 since
its a QoS NULL data frame.

Added this frame under exception route to bypass TCL with the help of FW.

Signed-off-by: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/dp_tx.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath11k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_tx.c
@@ -221,7 +221,8 @@ tcl_ring_sel:
 
 	switch (ti.encap_type) {
 	case HAL_TCL_ENCAP_TYPE_NATIVE_WIFI:
-		if (arvif->vif->offload_flags & IEEE80211_OFFLOAD_ENCAP_ENABLED)
+		if ((arvif->vif->offload_flags & IEEE80211_OFFLOAD_ENCAP_ENABLED) &&
+		    ieee80211_is_qos_nullfunc(hdr->frame_control))
 			is_diff_encap = true;
 		else
 			ath11k_dp_tx_encap_nwifi(skb);
